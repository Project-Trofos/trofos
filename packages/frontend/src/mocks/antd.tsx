import React from 'react';
import { Select as AntdSelect, DatePicker as AntdDatePicker } from 'antd';
import mockDayjs from 'dayjs';

// Mock certain antd component to simplify testing in React Testing Library
vi.mock('antd', async () => { 
  const antd: any = await vi.importActual('antd');
  antd.theme.defaultConfig.hashed = false;

  function DatePicker(props: React.ComponentProps<typeof AntdDatePicker>) {
    const { defaultValue, placeholder, disabled, id, onChange } = props;
    return (
      <input
        id={id}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        placeholder={placeholder}
        disabled={disabled}
        onChange={(e) => {
          if (onChange) {
            onChange(mockDayjs(e.target.value) as unknown as Parameters<typeof onChange>['0'], e.target.value);
          }
        }}
      />
    );
  }

  function RangePicker(props: React.ComponentProps<typeof AntdDatePicker.RangePicker>) {
    const { defaultValue, id, onChange } = props;
    return (
      <input
        id={id as string | undefined}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        onChange={(e) => {
          if (onChange) {
            onChange(
              [mockDayjs(e.target.value), mockDayjs(e.target.value)] as unknown as Parameters<typeof onChange>['0'],
              [e.target.value, e.target.value],
            );
          }
        }}
      />
    );
  }

  DatePicker.RangePicker = RangePicker;

  function Select(props: React.ComponentProps<typeof AntdSelect>) {
    const { mode, value, defaultValue, className, onChange, disabled, children, id } = props;

    const multiple = ['tags', 'multiple'].includes(mode ?? '');

    return (
      <select
        id={id}
        value={value as React.SelectHTMLAttributes<HTMLSelectElement>['value']}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLSelectElement>['defaultValue']}
        multiple={multiple}
        disabled={disabled}
        className={className}
        onChange={(e) => {
          if (onChange) {
            onChange(
              multiple ? Array.from(e.target.selectedOptions).map((option) => option.value) : e.target.value,
              {},
            );
          }
        }}
      >
        {children}
      </select>
    );
  }

  function Option({ children, ...otherProps }: { children: React.ReactNode }) {
    return <option {...otherProps}>{children}</option>;
  }
  function OptGroup({ children, ...otherProps }: { children: React.ReactNode }) {
    return <optgroup {...otherProps}>{children}</optgroup>;
  }

  Select.Option = Option;
  Select.OptGroup = OptGroup;

  return { ...antd, Select, DatePicker };
});
