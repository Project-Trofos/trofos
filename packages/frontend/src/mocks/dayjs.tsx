const dayjsMock = {
  year: () => '2022',
};

export default dayjsMock;
