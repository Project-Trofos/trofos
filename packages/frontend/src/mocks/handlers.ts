/* eslint-disable import/no-extraneous-dependencies */
import { rest } from 'msw';
import { CourseData } from '../api/types';

const BASE_URL = process.env.REACT_APP_BACKEND_BASE_URL || 'http://localhost:3001';
const NUSMODS_URL = 'https://api.nusmods.com/v2/2022-2023/moduleList.json';

export const MSW_COURSE: CourseData = {
  id: 1,
  code: 'CS3203',
  cname: 'Software Engineering Project',
  startYear: 2022,
  endYear: 2022,
  startSem: 1,
  endSem: 1,
  description: null,
  milestones: [
    {
      id: 1,
      name: 'Milestone 1',
      course_id: 1,
      start_date: '2023-12-31',
      deadline: '2023-01-09',
      created_at: '2022-12-18',
    },
    {
      id: 2,
      name: 'Milestone 2',
      course_id: 1,
      start_date: '2023-01-10',
      deadline: '2023-01-19',
      created_at: '2022-12-18',
    },
    {
      id: 3,
      name: 'Milestone 3',
      course_id: 1,
      start_date: '2023-01-20',
      deadline: '2023-01-30',
      created_at: '2022-12-18',
    },
  ],
  users: [],
  public: false,
  created_at: '2022-09-14T03:33:34.960Z',
};

const PROJECT = {
  id: 1,
  pname: 'project1',
  pkey: null,
  description: 'project1_description',
  course_id: '1',
  public: false,
  created_at: '2022-09-15T01:58:01.735Z',
  course: MSW_COURSE,
};

const NUSMODS_MODULE = {
  moduleCode: 'EL1101E',
  title: 'The Nature of Language',
  semesters: [1, 2],
};

const handlers = [
  // Handles GET on /project
  rest.get(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([PROJECT])))),

  // Handles GET on /course
  rest.get(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([MSW_COURSE])))),

  // Handles GET to NUSMODS
  rest.get(NUSMODS_URL, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([NUSMODS_MODULE])))),

  // Handles POST on /project
  rest.post(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/project`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  rest.post(`${BASE_URL}/course/:courseId/milestone`, (req, res, ctx) =>
    res(ctx.status(200), ctx.body(JSON.stringify({}))),
  ),
];

export default handlers;
