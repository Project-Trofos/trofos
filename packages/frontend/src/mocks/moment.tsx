const momentMock = {
  year: () => '2022',
};

export default momentMock;
