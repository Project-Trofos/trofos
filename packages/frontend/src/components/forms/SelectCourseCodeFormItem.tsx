import { Form, Select } from 'antd';
import React from 'react';

// Renders form elements to select course code from a list of options
export default function SelectCourseCodeFormItem({
  courseOptions,
  searchCourseKeyword,
  setSearchCourseKeyword,
}: {
  courseOptions: JSX.Element[],
  searchCourseKeyword: string,
  setSearchCourseKeyword: (value: string) => void,
}): JSX.Element {
  return (
    <Form.Item label="Course" name="courseCode" rules={[{ required: true, message: 'Please select course!' }]}>
      <Select
        showSearch
        value={searchCourseKeyword}
        onChange={(value) => setSearchCourseKeyword(value)}
        placeholder="Search to Select"
        optionFilterProp="children"
        filterOption={(input, option) => {
          if (!option) {
            return false;
          }
          return (option.key as string).toLowerCase().includes(input.toLowerCase());
        }}
      >
        {courseOptions}
      </Select>
    </Form.Item>
  );
}
