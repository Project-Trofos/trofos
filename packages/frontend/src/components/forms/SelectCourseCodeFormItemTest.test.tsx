import React from 'react';
import '../../mocks/antd';
import { render, screen } from '@testing-library/react';
import { Select } from 'antd';
import FormWrapper from './__testutils__/FormRenderHelper';
import SelectCourseCodeFormItem from './SelectCourseCodeFormItem';

describe('SelectCourseCodeFormItem test', () => {
  function renderForm() {
    return render(
      <FormWrapper>
        <SelectCourseCodeFormItem
          courseOptions={[
            <Select.Option key="CS3203" value="CS3203">
              CS3203 Software Engineering Project
            </Select.Option>,
            <Select.Option key="CS1010" value="CS1010">
              CS1010 Programming Methodology
            </Select.Option>,
          ]}
        />
      </FormWrapper>,
    );
  }

  describe('when rendering', () => {
    it('should have correct fields', () => {
      const { baseElement } = renderForm();
      expect(baseElement).toMatchSnapshot();

      const select = screen.getByLabelText(/Course/i);

      expect(select).toBeInTheDocument();

      screen.getByText('CS1010', { exact: false });
      screen.getByText('CS3203', { exact: false });
    });
  });
});
