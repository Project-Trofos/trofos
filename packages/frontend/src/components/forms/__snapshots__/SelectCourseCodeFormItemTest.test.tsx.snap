// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectCourseCodeFormItem test > when rendering > should have correct fields 1`] = `
<body>
  <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-form-item"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="courseCode"
              title="Course"
            >
              Course
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <select
                  id="courseCode"
                >
                  <option
                    value="CS3203"
                  >
                    CS3203 Software Engineering Project
                  </option>
                  <option
                    value="CS1010"
                  >
                    CS1010 Programming Methodology
                  </option>
                </select>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="submit"
      >
        <span>
          submit
        </span>
      </button>
    </form>
  </div>
</body>
`;
